### PR TITLE
Tighten the NM definition.

### LINKED ISSUE #177: Value of NM tag when ambiguous bases are involved
The definition of the NM tag is 
  > Edit distance to the reference, including ambiguous bases but excluding clipping.

The Tags file doesn't edify much more. It find it unclear, in particular, what does "including ambiguous codes" mean? 

I can understand it in two different ways:
A:
> "taking into account that an ambiguous code is always an edit distance away from anything" 

B: 
>"if a read base matches the pattern of the reference the edit distance is 0, else 1"

Here's a table of examples with the two possible interpretations:
 
\#  | Read | Ref | NM (A) | NM (B) | comments
----|----|----|----|----|---
1| A | A | 0| 0| clearly..
2| A| a | 0| 0| since bases as case-insensitive
3| A |G |1 | 1| clearly
4| A | g| 1 | 1 | base-insensitive
5| A| N | 1 | 0|
6| N| N | 1 | 0|
7 | N |A | 1| 1 | Since the read is less specific than the reference? not sure
8 | A| M |1 | 0 | M is A or C
9 | M| A | 1 | 1 | again, not sure
10 | M|R | 1 | 1| R is A or G...not sure here
10|M |M| 1 |0|

A discussion would be great, and when we reach an understanding, I'll be happy to submit a PR for the spec as well.

### PR DESCRIPTION
**Pull request by jkbonfield**
This is something that doesn't seem to correct match any existing
tool out there with regard to the full gamut of corner cases, but in
that regard it has no favourites!

Fixes #177

### CODE REVIEW

**Comment by jkbonfield** [comment]
For added fun I explored the "=" base type, meaning "matches reference".  You may think this wouldn't increase NM, being a match and all...  You'd be wrong. ;-)

Samtools calmd worked OK, but the CRAM code and Picard's NM calculation treated = as mismatch.  I didn't dare test putting = in reference, but that's illegal so out of scope anyway.   We could possibly go the extra mile and describe how to deal with illegal data, such as "Z" in either sequence or reference being treated as if they are "N", but I think this ought to be self explanatory and not necessary given we explicitly state which characters can NOT give rise to a mismatch.

File: SAMtags.tex [hunk 1]
```diff
@@ -190,7 +190,12 @@ \subsection{Additional Template and Mapping data}
 Number of reported alignments that contains the query in the current record.
 
 \item[NM:i:\tagvalue{}]
-Edit distance to the reference, including ambiguous bases but excluding clipping.
+Number of mismatches between the sequence and reference, counting only pure (case-insensitive) A, C, G and T bases in sequence and reference as potential matches with everything else being a mismatch.
```
[the remark this answers is not in the document]
Comment (jkbonfield) [reply SAMtags.tex]: Granted it is a bit wordy, but hard to be succinct with all the corner cases.

I like your wording, but will wait for others to respond before editing.  Thanks.

**Comment by EvanTheB** [comment]
would love to see this go through, I am working around the differences between tools for some work on CRAM. It would be much easier if tools had 'bugs' instead of 'different interpretations of the spec'. Thanks for your work

**Comment by jkbonfield** [comment]
With the latest htslib you can just force cram to store NM and MD verbatim, which in turn means decode just returns whatever it had originally.  It costs space, but it's a workaround if the same interpretation is necessary for you.

See store_md and store_nm in the samtools man page: http://www.htslib.org/doc/samtools.html#GLOBAL_OPTIONS

An example:
```
samtools view -O cram,store_md=1,store_nm=1 -o aln.cram aln.bam
samtools view --input-fmt cram,decode_md=0 -o aln.new.bam aln.cram
```

**Comment by yfarjoun** [comment]
Add a note in the log part to state that the definition has been tightened. otherwise, I think that this is fine. 

👍

**Comment by jkbonfield** [comment]
Added a log message, rebased and squashed.

**Comment by jkbonfield** [comment]
@jmarshall?  2 out of 3 so far ;-)

**Comment by jmarshall** [comment]
Thanks for filling in the tag value. This is :+1: Make It So from me, with some typographical miscellanea — which I've pushed as a commit as that's easier than describing it all.

Oxford commas, another comma elsewhere for clarity, etc. Please pick and choose which if any miscellanea you like and squash it in… otherwise thanks and please merge your NM!

(Me, I prefer Oxford commas (so `only (case-insensitive) A, C, G, and T bases` and `‘N’, ‘P’, ‘S’, and ‘H’ CIGAR operations`). Don't really mind whether you act on this one or not, but note the spec does tend to use them.)

**Comment by jkbonfield** [comment]
Oxford commas can disambiguate things in some cases and don't look too bad with long words (but it's not used in .g. "For example, it allows a quick way of checking that references named `1', `Chr1' and `chr1' in different files are in fact the same").  I'm less convinced with shorter words.  To me it just looks wrong on A, C, G, and T.

Thanks for the other changes though.  I'll get it squashed and merged.

**Comment by jmarshall** [comment]
I'll be sure to fix that “checking that references … are in fact the same” one :smile:

To me, “A, C, G and T” looks like a refreshing drink.

**Comment by jkbonfield** [comment]
Fixed by 4eacedebb9bf0e5f67375b76a35a2aa71b6800f1
